sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,498 @@
+    instructions during LLVM->SPIR-V translation.
+- Based on some attributes (which could be defined later) we may want to avoid
+  listing some functions as imported ones
+  - This is needed to have possibility to call device-specific builtins not
+    starting with `__` by forward-declaring them in DPC++ code
+
+**NOTE**: If device code split is enabled, imports collection is performed after
+split and it is performed on separated images.
+
+All collected information is attached to a device image via properties
+mechanism.
+
+Each device image is supplied with an array of property sets:
+```
+struct pi_device_binary_struct {
+...
+  // Array of property sets
+  pi_device_binary_property_set PropertySetsBegin;
+  pi_device_binary_property_set PropertySetsEnd;
+};
+
+```
+Each property set is represented by the following struct:
+```
+// Named array of properties.
+struct _pi_device_binary_property_set_struct {
+  char *Name;                                // the name
+  pi_device_binary_property PropertiesBegin; // array start
+  pi_device_binary_property PropertiesEnd;   // array end
+};
+```
+It contains name of property set and array of properties. Each property is
+represented by the following struct:
+```
+struct _pi_device_binary_property_struct {
+  char *Name;       // null-terminated property name
+  void *ValAddr;    // address of property value
+  uint32_t Type;    // _pi_property_type
+  uint64_t ValSize; // size of property value in bytes
+};
+```
+
+List of imported symbols is represented as a single property set with name
+`SYCL/imported symbols` recorded in the `Name` field of property set.
+Each property in this set holds name of the particular imported symbol recorded
+in the `Name` field of the property.
+List of exported symbols is represented in the same way, except the
+corresponding set has the name `SYCL/exported symbols`.
+
+### DPC++ runtime changes
+
+DPC++ RT performs *device images collection* task by grouping all device
+images required to execute a kernel based on the list of exports/imports and
+links them together using PI API.
+This native device image is then added to the cache to avoid symbol resolution,
+compilation, and linking for any future attempts to invoke kernels from this
+device image.
+
+#### Device images collection and linking
+
+Device images collection and linking is performed by DPC++ Runtime class named
+`ProgramManager`.
+
+When the program manager gets a request to JIT compile a device image(program)
+it examines its list of imported symbols and finds device images which exports
+those symbols, then links requested device image and images found together.
+
+All needed device images are found by iterating through all available OS modules
+without predefined order and searching for first unresolved symbol in list of
+imports of target device image. Once device image that contains first symbol is
+met, remaining exported symbols are checked in found image. If
```

Comment:
Should we say anything about types of device images more explicitly? I guess there is no sense in inspecting device images with native code if the application only contains a device image with SPIR-V

Reply:
Done.